File: python3/add_one_row_to_tree.py
from .model import TreeNode

"""
Space   : O(n)
Time    : O(n)
"""


class Solution:
    def addOneRow(self, root: Optional[TreeNode], val: int, depth: int) -> Optional[TreeNode]:
        if depth == 1:
            return TreeNode(val, root)

        stack = [root]

        level = 2
        while stack:
            temp = []
            for node in stack:
                if level == depth:
                    left = node.left
                    right = node.right
                    node.left = TreeNode(val)
                    node.left.left = left
                    node.right = TreeNode(val)
                    node.right.right = right

                if node.left:
                    temp.append(node.left)
                if node.right:
                    temp.append(node.right)

            if level == depth:
                break
            stack = temp
            level += 1

        return root

File: golang/add_one_row_to_tree.go
package main

// Space	: O(n)
// Time		: O(n)

import (
	"golang/model"
)

func addOneRow(root *model.TreeNode, v int, d int) *model.TreeNode {
	if d == 1 {
		ans := &model.TreeNode{
			Val:  v,
			Left: root,
		}
		return ans
	}

	level := 1
	stack := []*model.TreeNode{root}

	for len(stack) > 0 {
		temp := []*model.TreeNode{}
		if level+1 == d {
			for _, node := range stack {
				if node.Left != nil {
					l := node.Left
					node.Left = &model.TreeNode{
						Val:  v,
						Left: l,
					}
				} else {
					node.Left = &model.TreeNode{
						Val: v,
					}
				}

				if node.Right != nil {
					r := node.Right
					node.Right = &model.TreeNode{
						Val:   v,
						Right: r,
					}
				} else {
					node.Right = &model.TreeNode{
						Val: v,
					}
				}
			}
			break
		}
		for _, node := range stack {
			if node.Left != nil {
				temp = append(temp, node.Left)
			}
			if node.Right != nil {
				temp = append(temp, node.Right)
			}
		}

		stack = temp
		level++
	}

	return root
}
